File: crates/rolldown_testing/src/workspace.rs
use std::path::PathBuf;

/// Get the absolute path to the root of the workspace/repository.
/// The root is always the directory containing the root `Cargo.toml`, `package.json`, `pnpm-workspace.yaml` etc.
pub fn root_dir() -> PathBuf {
  PathBuf::from(env!("WORKSPACE_DIR"))
}

pub fn crate_dir(crate_name: &str) -> PathBuf {
  let root = root_dir();
  root.join("crates").join(crate_name)
}

File: packages/testing/src/workspace.ts
import nodeAssert from 'node:assert';
import nodeFs from 'node:fs';
import nodePath from 'node:path';

/**
 * @description
 * - Get the absolute path to the root of the workspace. The root is always the directory containing the root `Cargo.toml`, `package.json`, `pnpm-workspace.yaml` etc.
 * - `workspaceRoot('packages')` equals to `path.resolve(workspaceRoot(), 'packages')`
 */
export function workspaceRoot(...joined: string[]) {
  return nodePath.resolve(import.meta.dirname, '../../..', ...joined);
}

nodeAssert(
  nodeFs.existsSync(workspaceRoot('pnpm-workspace.yaml')),
  `${workspaceRoot('pnpm-workspace.yaml')} does not exist`,
);
